File: app/src/main/java/com/example/data/seed/DatabaseSeeder.kt
package com.example.data.seed

import com.example.domain.model.Account
import com.example.domain.model.AccountType
import com.example.domain.model.CurrencyUtils
import com.example.domain.model.Voucher
import com.example.domain.model.VoucherLine
import com.example.repository.AccountingRepository

object DatabaseSeeder {
    suspend fun seedInitialData(repository: AccountingRepository) {
        val accounts = repository.getAllAccounts()
        if (accounts.isNotEmpty()) {
            return
        }

        // ==========================================
        // 1. LEVEL 1: MAJOR HEADS (No Parent)
        // ==========================================
        val l1Assets = repository.createAccount(
            Account(code = "1000", name = "Assets", parentId = null, level = 1, accountType = AccountType.ASSET, description = "Economic resources owned by the business")
        ).getOrThrow()

        val l1Liabilities = repository.createAccount(
            Account(code = "2000", name = "Liabilities", parentId = null, level = 1, accountType = AccountType.LIABILITY, description = "Present obligations of the enterprise")
        ).getOrThrow()

        val l1Equity = repository.createAccount(
            Account(code = "3000", name = "Equity", parentId = null, level = 1, accountType = AccountType.EQUITY, description = "Residual interest in the assets of the enterprise")
        ).getOrThrow()

        val l1Revenue = repository.createAccount(
            Account(code = "4000", name = "Revenue", parentId = null, level = 1, accountType = AccountType.REVENUE, description = "Gross inflow of economic benefits")
        ).getOrThrow()

        val l1Expenses = repository.createAccount(
            Account(code = "5000", name = "Expenses", parentId = null, level = 1, accountType = AccountType.EXPENSE, description = "Decreases in economic benefits during the accounting period")
        ).getOrThrow()

        // ==========================================
        // 2. LEVEL 2: SUB-HEADS (Parent: Level 1)
        // ==========================================
        val l2CurrentAssets = repository.createAccount(
            Account(code = "1100", name = "Current Assets", parentId = l1Assets.id, level = 2, accountType = AccountType.ASSET, description = "Short-term operating assets")
        ).getOrThrow()

        val l2NonCurrentAssets = repository.createAccount(
            Account(code = "1200", name = "Non-Current Assets", parentId = l1Assets.id, level = 2, accountType = AccountType.ASSET, description = "Long-term fixed equipment and property")
        ).getOrThrow()

        val l2CurrentLiabilities = repository.createAccount(
            Account(code = "2100", name = "Current Liabilities", parentId = l1Liabilities.id, level = 2, accountType = AccountType.LIABILITY, description = "Short-term debts and payables")
        ).getOrThrow()

        val l2OwnersEquity = repository.createAccount(
            Account(code = "3100", name = "Owner's Equity", parentId = l1Equity.id, level = 2, accountType = AccountType.EQUITY, description = "Contributed capital and reserves")
        ).getOrThrow()

        val l2OperatingRevenue = repository.createAccount(
            Account(code = "4100", name = "Operating Revenue", parentId = l1Revenue.id, level = 2, accountType = AccountType.REVENUE, description = "Core service and sales income")
        ).getOrThrow()

        val l2OperatingExpenses = repository.createAccount(
            Account(code = "5100", name = "Operating Expenses", parentId = l1Expenses.id, level = 2, accountType = AccountType.EXPENSE, description = "Day-to-day business operational expenses")
        ).getOrThrow()

        // ==========================================
        // 3. LEVEL 3: CONTROL GROUPS (Parent: Level 2)
        // ==========================================
        val l3CashAndBank = repository.createAccount(
            Account(code = "1110", name = "Cash & Bank", parentId = l2CurrentAssets.id, level = 3, accountType = AccountType.ASSET, description = "Liquid cash funds and bank holdings")
        ).getOrThrow()

        val l3Receivables = repository.createAccount(
            Account(code = "1120", name = "Accounts Receivable", parentId = l2CurrentAssets.id, level = 3, accountType = AccountType.ASSET, description = "Customer balances due")
        ).getOrThrow()

        val l3OfficeEquipment = repository.createAccount(
            Account(code = "1210", name = "Office Equipment", parentId = l2NonCurrentAssets.id, level = 3, accountType = AccountType.ASSET, description = "IT hardware and office machinery")
        ).getOrThrow()

        val l3Payables = repository.createAccount(
            Account(code = "2110", name = "Accounts Payable", parentId = l2CurrentLiabilities.id, level = 3, accountType = AccountType.LIABILITY, description = "Trade creditors and short-term payables")
        ).getOrThrow()

        val l3Capital = repository.createAccount(
            Account(code = "3110", name = "Capital Accounts", parentId = l2OwnersEquity.id, level = 3, accountType = AccountType.EQUITY, description = "Partner and shareholder capital")
        ).getOrThrow()

        val l3ServiceRevenue = repository.createAccount(
            Account(code = "4110", name = "Professional Services", parentId = l2OperatingRevenue.id, level = 3, accountType = AccountType.REVENUE, description = "Professional client fees")
        ).getOrThrow()

        val l3AdminExpenses = repository.createAccount(
            Account(code = "5110", name = "Administrative Expenses", parentId = l2OperatingExpenses.id, level = 3, accountType = AccountType.EXPENSE, description = "Office and general admin expenditure")
        ).getOrThrow()

        val l3UtilityExpenses = repository.createAccount(
            Account(code = "5120", name = "Utilities & Facility", parentId = l2OperatingExpenses.id, level = 3, accountType = AccountType.EXPENSE, description = "Power, internet, water and lease costs")
        ).getOrThrow()

        // ==========================================
        // 4. LEVEL 4: POSTING ACCOUNTS (Parent: Level 3)
        // ==========================================
        val cashInHand = repository.createAccount(
            Account(code = "111001", name = "Cash in Hand", parentId = l3CashAndBank.id, level = 4, accountType = AccountType.ASSET, description = "Main physical cash register")
        ).getOrThrow()

        val pettyCash = repository.createAccount(
            Account(code = "111002", name = "Petty Cash", parentId = l3CashAndBank.id, level = 4, accountType = AccountType.ASSET, description = "Imprest petty cash float")
        ).getOrThrow()

        val bankAccount = repository.createAccount(
            Account(code = "111003", name = "Main Bank Account", parentId = l3CashAndBank.id, level = 4, accountType = AccountType.ASSET, description = "Operating commercial bank account")
        ).getOrThrow()

        val tradeDebtors = repository.createAccount(
            Account(code = "112001", name = "Trade Debtors", parentId = l3Receivables.id, level = 4, accountType = AccountType.ASSET, description = "Accounts receivable trade customers")
        ).getOrThrow()

        val officeComputers = repository.createAccount(
            Account(code = "121001", name = "Computers & IT Hardware", parentId = l3OfficeEquipment.id, level = 4, accountType = AccountType.ASSET, description = "Laptops and desktop workstations")
        ).getOrThrow()

        val tradeCreditors = repository.createAccount(
            Account(code = "211001", name = "Trade Creditors", parentId = l3Payables.id, level = 4, accountType = AccountType.LIABILITY, description = "Commercial suppliers payables")
        ).getOrThrow()

        val ownerCapital = repository.createAccount(
            Account(code = "311001", name = "Owner Capital Contribution", parentId = l3Capital.id, level = 4, accountType = AccountType.EQUITY, description = "Initial equity injection")
        ).getOrThrow()

        val consultingRevenue = repository.createAccount(
            Account(code = "411001", name = "Technical Consulting Revenue", parentId = l3ServiceRevenue.id, level = 4, accountType = AccountType.REVENUE, description = "Technical consultancy billings")
        ).getOrThrow()

        val officeExpense = repository.createAccount(
            Account(code = "511001", name = "Office Expense", parentId = l3AdminExpenses.id, level = 4, accountType = AccountType.EXPENSE, description = "Office supplies, stationery and printer costs")
        ).getOrThrow()

        val travelExpense = repository.createAccount(
            Account(code = "511002", name = "Travel & Local Transport", parentId = l3AdminExpenses.id, level = 4, accountType = AccountType.EXPENSE, description = "Local client visits and taxi receipts")
        ).getOrThrow()

        val electricityExpense = repository.createAccount(
            Account(code = "512001", name = "Electricity & Power", parentId = l3UtilityExpenses.id, level = 4, accountType = AccountType.EXPENSE, description = "Monthly office electricity billing")
        ).getOrThrow()

        val internetExpense = repository.createAccount(
            Account(code = "512002", name = "Internet & Communication", parentId = l3UtilityExpenses.id, level = 4, accountType = AccountType.EXPENSE, description = "High-speed broadband and telephony")
        ).getOrThrow()

        val rentExpense = repository.createAccount(
            Account(code = "512003", name = "Office Rent Expense", parentId = l3UtilityExpenses.id, level = 4, accountType = AccountType.EXPENSE, description = "Commercial premises rent")
        ).getOrThrow()

        // ==========================================
        // 5. SAMPLE CP VOUCHER (CP-000001)
        // ==========================================
        val initialVoucher = Voucher(
            voucherNumber = "CP-000001",
            voucherType = "CP",
            voucherDate = System.currentTimeMillis() - (86400000L * 2), // 2 days ago
            reference = "RCPT-2026-001",
            narration = "Payment for Office Supplies and Stationery",
            lines = listOf(
                VoucherLine(
                    accountId = officeExpense.id,
                    description = "Office Stationery and Supplies",
                    debitCents = CurrencyUtils.doubleToCents(5000.0), // 5,000.00
                    creditCents = 0L
                ),
                VoucherLine(
                    accountId = cashInHand.id,
                    description = "Cash Disbursed from Main Cash",
                    debitCents = 0L,
                    creditCents = CurrencyUtils.doubleToCents(5000.0) // 5,000.00
                )
            )
        )

        val createdVoucher = repository.createVoucher(initialVoucher).getOrThrow()
        repository.postVoucher(createdVoucher.id).getOrThrow()
    }
}

File: backend/src/Accounting.Infrastructure/Data/DatabaseSeeder.cs
using Accounting.Domain.Entities;
using Microsoft.EntityFrameworkCore;

namespace Accounting.Infrastructure.Data;

public static class DatabaseSeeder
{
    public static async Task SeedAsync(AccountingDbContext context)
    {
        if (await context.Accounts.AnyAsync())
        {
            return; // Already seeded
        }

        // --- LEVEL 1: Major Heads ---
        var l1Assets = new Account { Code = "1000", Name = "Assets", Level = 1, ParentId = null, AccountType = AccountType.Asset };
        var l1Liabilities = new Account { Code = "2000", Name = "Liabilities", Level = 1, ParentId = null, AccountType = AccountType.Liability };
        var l1Equity = new Account { Code = "3000", Name = "Equity", Level = 1, ParentId = null, AccountType = AccountType.Equity };
        var l1Revenue = new Account { Code = "4000", Name = "Revenue", Level = 1, ParentId = null, AccountType = AccountType.Revenue };
        var l1Expenses = new Account { Code = "5000", Name = "Expenses", Level = 1, ParentId = null, AccountType = AccountType.Expense };

        await context.Accounts.AddRangeAsync(l1Assets, l1Liabilities, l1Equity, l1Revenue, l1Expenses);
        await context.SaveChangesAsync();

        // --- LEVEL 2: Sub-heads ---
        var l2CurrentAssets = new Account { Code = "1100", Name = "Current Assets", Level = 2, ParentId = l1Assets.Id, AccountType = AccountType.Asset };
        var l2FixedAssets = new Account { Code = "1200", Name = "Fixed Assets", Level = 2, ParentId = l1Assets.Id, AccountType = AccountType.Asset };
        var l2CurrentLiab = new Account { Code = "2100", Name = "Current Liabilities", Level = 2, ParentId = l1Liabilities.Id, AccountType = AccountType.Liability };
        var l2Capital = new Account { Code = "3100", Name = "Capital & Reserves", Level = 2, ParentId = l1Equity.Id, AccountType = AccountType.Equity };
        var l2OperatingRev = new Account { Code = "4100", Name = "Operating Revenue", Level = 2, ParentId = l1Revenue.Id, AccountType = AccountType.Revenue };
        var l2AdminExp = new Account { Code = "5100", Name = "Administrative Expenses", Level = 2, ParentId = l1Expenses.Id, AccountType = AccountType.Expense };

        await context.Accounts.AddRangeAsync(l2CurrentAssets, l2FixedAssets, l2CurrentLiab, l2Capital, l2OperatingRev, l2AdminExp);
        await context.SaveChangesAsync();

        // --- LEVEL 3: Control Groups ---
        var l3CashBank = new Account { Code = "1110", Name = "Cash and Bank Balances", Level = 3, ParentId = l2CurrentAssets.Id, AccountType = AccountType.Asset };
        var l3Receivables = new Account { Code = "1120", Name = "Accounts Receivable", Level = 3, ParentId = l2CurrentAssets.Id, AccountType = AccountType.Asset };
        var l3OfficeEquipGroup = new Account { Code = "1210", Name = "Office Equipment", Level = 3, ParentId = l2FixedAssets.Id, AccountType = AccountType.Asset };
        var l3Payables = new Account { Code = "2110", Name = "Accounts Payable", Level = 3, ParentId = l2CurrentLiab.Id, AccountType = AccountType.Liability };
        var l3OwnersEquity = new Account { Code = "3110", Name = "Owner's Equity", Level = 3, ParentId = l2Capital.Id, AccountType = AccountType.Equity };
        var l3Sales = new Account { Code = "4110", Name = "Sales Income", Level = 3, ParentId = l2OperatingRev.Id, AccountType = AccountType.Revenue };
        var l3OfficeOps = new Account { Code = "5110", Name = "Office Operations", Level = 3, ParentId = l2AdminExp.Id, AccountType = AccountType.Expense };
        var l3Utilities = new Account { Code = "5120", Name = "Utility Expenses", Level = 3, ParentId = l2AdminExp.Id, AccountType = AccountType.Expense };

        await context.Accounts.AddRangeAsync(l3CashBank, l3Receivables, l3OfficeEquipGroup, l3Payables, l3OwnersEquity, l3Sales, l3OfficeOps, l3Utilities);
        await context.SaveChangesAsync();

        // --- LEVEL 4: Posting Accounts ---
        var l4CashInHand = new Account { Code = "111001", Name = "Cash in Hand", Level = 4, ParentId = l3CashBank.Id, AccountType = AccountType.Asset };
        var l4MainBank = new Account { Code = "111002", Name = "Main Commercial Bank", Level = 4, ParentId = l3CashBank.Id, AccountType = AccountType.Asset };
        var l4Computers = new Account { Code = "121001", Name = "Computers & Laptops", Level = 4, ParentId = l3OfficeEquipGroup.Id, AccountType = AccountType.Asset };
        var l4TradeCreditors = new Account { Code = "211001", Name = "Trade Creditors Control", Level = 4, ParentId = l3Payables.Id, AccountType = AccountType.Liability };
        var l4ShareCapital = new Account { Code = "311001", Name = "Share Capital", Level = 4, ParentId = l3OwnersEquity.Id, AccountType = AccountType.Equity };
        var l4SoftwareConsulting = new Account { Code = "411001", Name = "Software Consulting Revenue", Level = 4, ParentId = l3Sales.Id, AccountType = AccountType.Revenue };
        var l4RentExpense = new Account { Code = "511001", Name = "Office Rent Expense", Level = 4, ParentId = l3OfficeOps.Id, AccountType = AccountType.Expense };
        var l4Supplies = new Account { Code = "511002", Name = "Office Stationery & Supplies", Level = 4, ParentId = l3OfficeOps.Id, AccountType = AccountType.Expense };
        var l4Electricity = new Account { Code = "512001", Name = "Electricity & Power", Level = 4, ParentId = l3Utilities.Id, AccountType = AccountType.Expense };
        var l4Internet = new Account { Code = "512002", Name = "Internet & Communication", Level = 4, ParentId = l3Utilities.Id, AccountType = AccountType.Expense };

        await context.Accounts.AddRangeAsync(
            l4CashInHand, l4MainBank, l4Computers, l4TradeCreditors,
            l4ShareCapital, l4SoftwareConsulting, l4RentExpense, l4Supplies,
            l4Electricity, l4Internet
        );
        await context.SaveChangesAsync();

        // --- Seed Sample Posted CP Voucher (Cash Payment) ---
        var sampleDate = DateTime.UtcNow.AddDays(-2);
        var voucher = new Voucher
        {
            VoucherNumber = "CP-000001",
            VoucherType = "CP",
            VoucherDate = sampleDate,
            Reference = "INV-2026-001",
            Narration = "Payment of office rent and utility charges via cash",
            TotalDebit = 5000.00m,
            TotalCredit = 5000.00m,
            IsPosted = true,
            PostedAtUtc = sampleDate.AddMinutes(5),
            Lines = new List<VoucherLine>
            {
                new() { AccountId = l4RentExpense.Id, Description = "Office rent for current month", Debit = 3500.00m, Credit = 0m },
                new() { AccountId = l4Electricity.Id, Description = "Electricity utility bill", Debit = 1500.00m, Credit = 0m },
                new() { AccountId = l4CashInHand.Id, Description = "Cash disbursed from drawer", Debit = 0m, Credit = 5000.00m }
            }
        };

        await context.Vouchers.AddAsync(voucher);
        await context.SaveChangesAsync();

        // Ledger Transactions for CP-000001
        var ledgerEntries = new List<LedgerTransaction>
        {
            new() { Date = sampleDate, VoucherId = voucher.Id, VoucherNumber = voucher.VoucherNumber, VoucherType = "CP", AccountId = l4RentExpense.Id, Description = "Office rent for current month", Debit = 3500.00m, Credit = 0m },
            new() { Date = sampleDate, VoucherId = voucher.Id, VoucherNumber = voucher.VoucherNumber, VoucherType = "CP", AccountId = l4Electricity.Id, Description = "Electricity utility bill", Debit = 1500.00m, Credit = 0m },
            new() { Date = sampleDate, VoucherId = voucher.Id, VoucherNumber = voucher.VoucherNumber, VoucherType = "CP", AccountId = l4CashInHand.Id, Description = "Cash disbursed from drawer", Debit = 0m, Credit = 5000.00m }
        };

        await context.LedgerTransactions.AddRangeAsync(ledgerEntries);
        await context.SaveChangesAsync();
    }
}
